source/posix/rw_lock.c

```diff
@@ -12,48 +12,103 @@
  * express or implied. See the License for the specific language governing
  * permissions and limitations under the License.
  */
-
-#include <aws/common/atomics.h>
+#include <aws/common/config.h>
 #include <aws/common/rw_lock.h>
 
 #include <aws/common/posix/common.inl>
 
+#include <assert.h>
+
+#if AWS_SIZEOF_VOID_P == 8
+typedef int64_t reader_count_t;
+static const int64_t s_max_readers = INT64_MAX;
+#elif AWS_SIZEOF_VOID_P == 4
+typedef int32_t reader_count_t;
+static const int32_t s_max_readers = INT32_MAX;
+#else
+#    error Unsupported pointer size
+#endif
+
 int aws_rw_lock_init(struct aws_rw_lock *lock) {
 
-    return aws_private_convert_and_raise_error_code(pthread_rwlock_init(&lock->lock_handle, NULL));
+    aws_atomic_init_int(&lock->readers, 0);
+    aws_atomic_init_int(&lock->holdouts, 0);
+    if (aws_semaphore_init(&lock->reader_sem, 0, s_max_readers)) {
+        return AWS_OP_ERR;
+    }
+    if (aws_semaphore_init(&lock->writer_sem, 0, 1)) {
+        aws_semaphore_clean_up(&lock->reader_sem);
+        return AWS_OP_ERR;
+    }
+    if (aws_mutex_init(&lock->writer_lock)) {
+        aws_semaphore_clean_up(&lock->reader_sem);
+        aws_semaphore_clean_up(&lock->writer_sem);
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_ERR;
 }
 
 void aws_rw_lock_clean_up(struct aws_rw_lock *lock) {
 
-    pthread_rwlock_destroy(&lock->lock_handle);
+    aws_semaphore_clean_up(&lock->reader_sem);
+    aws_semaphore_clean_up(&lock->writer_sem);
+    aws_mutex_clean_up(&lock->writer_lock);
 }
 
 int aws_rw_lock_rlock(struct aws_rw_lock *lock) {
 
-    return aws_private_convert_and_raise_error_code(pthread_rwlock_rdlock(&lock->lock_handle));
+    const reader_count_t num_readers = aws_atomic_fetch_add(&lock->readers, 1) + 1;
+    if (num_readers < 0) {
+        aws_semaphore_acquire(&lock->reader_sem);
+    }
+    assert(aws_atomic_load_int(&lock->readers) > 0);
```

Comment:
Problematic race condition:

* R: aws_rw_lock_rlock
* R: lock->readers += 1; num_readers = 1;
* W: aws_rw_lock_wlock
* W: lock->readers -= s_max_readers; lock->readers = 1 - INT64_MAX
* R: assert(lock->readers > 0) // assert fail

Reply:
Oh you're right. These asserts were useful in testing but I think I should remove them before more complex cases cause problems.

Reply:
This highlights the need for documentation; if the asserts are wrong, it suggests that your mental model when writing this code was wrong, and so a deeper analysis is needed to ensure that this is correct.

Have you built a TLA+ model for this rwlock protocol?

Reply:
Also, do you have a test that hammers this with concurrent operations?

Reply:
The rw_lock tests should hit this pretty hard.

As for the asserts, I simply added them to validate my understandings of how this code works in a single-threaded context. The implementation is identical to the one in the [C++ SDK](https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/utils/threading/ReaderWriterLock.cpp).

Reply:
The behavior of the code in a single-threaded context is not all that interesting for multithreaded code - and the implementation is far from identical.

Again, you're writing a nontrivial synchronization protocol here, what have we done to document the assumptions and invariants of this protocol, and what have we done to verify correctness?

Reply:
Disagree, if you inline all C++ operators that are obviously not available here, the code should be functionally the same. If you spot any discrepancies, please call them out.

Please see docs added since your review.